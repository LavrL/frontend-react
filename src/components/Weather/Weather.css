.button-weather__orange {
    height: 30px;
    font-size: 14px;
    border-radius: 8px;
    outline: none;
    border: 1px solid #fca311;
    background-color: #fca311;
    background-image: linear-gradient(-245deg, rgba(255, 255, 255, 0.3) 0%, rgba(255, 255, 255, 0) 100%);
    width: 150px;
}
.weather-error {
    color: red;
}
.weather-block__center {
    text-align: center;
}
.weather__input-text {
    height: 18px;
    border-radius: 2px;
    border: none;
    margin-right: 10px;
    padding: 14px;
    font-size: 15px;
    outline: none;
}
.weather__forecast_number{
    color: orange;
}
.weather__forecast_title {
    color: orange;
    font-size: 20px;
}