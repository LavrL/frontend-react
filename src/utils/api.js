export const API = 'https://api.openweathermap.org/data/2.5/weather';

export const api1 = (city, country, REACT_APP_WEATHER_APPID) => {
    return fetch(`${API}?q=${city},${country}&units=metric&appid=${REACT_APP_WEATHER_APPID}`)
        .then(res => { return res.json() });

};
