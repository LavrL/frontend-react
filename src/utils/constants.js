export const DEFAULT_IMAGE = 'https://m.media-amazon.com/images/M/MV5BMTczNTI2ODUwOF5BMl5BanBnXkFtZTcwMTU0NTIzMw@@._V1_SX300.jpg';
export const MOVIE_API = "https://www.omdbapi.com/?s=batman&apikey=5d508a92";
export const REACT_APP_WEATHER_APPID = 'd416957097d97116bdbffca2895ada24'
