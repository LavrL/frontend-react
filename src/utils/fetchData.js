import axios from 'axios';

export const API = 'https://api.openweathermap.org/data/2.5/weather';
export const fetchData = async (city, country, REACT_APP_WEATHER_APPID) => {
    const url = `${API}?q=${city},${country}&units=metric&appid=${REACT_APP_WEATHER_APPID}`;

    return await axios.get(url);
};