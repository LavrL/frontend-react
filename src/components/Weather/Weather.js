import React, { Component } from 'react';
import { REACT_APP_WEATHER_APPID } from '../../utils/constants';
import InputComponent from '../InputComponent/InputComponent';

import './Weather.css';

const initiateState = {
    temp: '',
    maxTemp: '',
    minTemp: '',
    main: '',
    city: 'Riga',
    cityToShow: '',
    countryToShow: '',
    country: 'LV',
    showWeather: false,
    showError: false
}

class Weather extends Component {
    constructor(props) {
        super(props);
        this.state = initiateState;
    }

    onChange = (e) => {
        this.setState({
            [e.target.name]: e.target.value
        });
    }
    getWeather = () => {
        fetch(`https://api.openweathermap.org/data/2.5/weather?q=${this.state.city},${this.state.country}&units=metric&appid=${REACT_APP_WEATHER_APPID}`)
            .then((res) => { return res.json() })
            .then((result) => {
                /* istanbul ignore else*/
                if (!this.state.showWeather) {
                    this.setState({
                        showWeather: !this.state.showWeather
                    })
                }
                this.setState({
                    minTemp: result.main.temp_min,
                    maxTemp: result.main.temp_max,
                    temp: result.main.temp,
                    main: result.weather[0].description,
                    cityToShow: this.state.city,
                    countryToShow: this.state.country,
                    showError: false
                })
            })
            .catch(err => {
                console.log('Error = ', err);
                this.setState({
                    showWeather: false,
                    showError: true
                })
            })
    }

    render() {
        const { city, country } = this.state;
        return (
            <div className="weather-block__center">
                <InputComponent
                    type="text"
                    className="weather__input-text"
                    name="city"
                    place={city}
                    onChange={this.onChange}
                />
                <InputComponent
                    type="text"
                    className="weather__input-text"
                    name="country"
                    place={country}
                    onChange={this.onChange}
                />

                <button className="button-weather__orange"
                    onClick={this.getWeather}>Get Weather</button>
                <div> {this.state.showWeather && <React.Fragment>
                    <div>
                        <h2 className="weather__forecast_title">
                            {this.state.cityToShow}<span>, </span>
                            {this.state.countryToShow}
                        </h2>
                    </div>
                    <div>
                        <h3><p>Temperature</p>
                            <span className="weather__forecast_number">{this.state.temp}</span>
                        </h3>
                    </div>
                    <div>
                        <h3>
                            <p>Min. temp</p>
                            <span className="weather__forecast_number">{this.state.minTemp}</span>
                        </h3>
                        <h3>
                            <p>Max. temp</p>
                            <span className="weather__forecast_number">{this.state.maxTemp}</span>
                        </h3>
                    </div>
                    <div><h3>{this.state.main}</h3>
                    </div> </React.Fragment>}
                    {this.state.showError && <div className="weather-error">Incorrect City or Country</div>}
                </div>
            </div>
        )
    }
}

export default Weather;
