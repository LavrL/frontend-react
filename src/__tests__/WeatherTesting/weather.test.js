import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Weather from '../../components/Weather/Weather';
import renderer from 'react-test-renderer';
import { mount, configure, shallow } from 'enzyme';
import axios from 'axios';
import { REACT_APP_WEATHER_APPID } from '../../utils/constants';
import { fetchData } from '../../utils/fetchData';
import { api } from '../../utils/api';
import fetchMock from 'jest-fetch-mock';
//import InputComponent from '../../components/InputComponent/InputComponent';

jest.mock('axios');

configure({ adapter: new Adapter() });

describe('Weather testing', () => {

    let wrapper;
    const data = {
        result: {
            temp: '10',
            minTemp: '6',
            maxTemp: '12',
            description: 'rain'
        }
    };

    beforeAll(() => {
        wrapper = mount(<Weather />);
        fetchMock.resetMocks();
    });

    afterAll(() => {
        wrapper.unmount();
    });

    test('getWeather button click event', () => {
        const mockCallBack = jest.fn();

        const buttonGetWeather = shallow(<button onClick={mockCallBack} />);
        buttonGetWeather.getElement().props.onClick();
        expect(mockCallBack).toHaveBeenCalled();
    });

    test('should handle click correctly', async () => {
        fetch.mockResponseOnce(JSON.stringify({main : { temp_min : '123'} }));
        const buttonWrapper = shallow(<Weather />);

        expect(buttonWrapper.find('button.button-weather__orange').length).toBe(1);
        expect(wrapper.state().showWeather).toBe(false);

        expect(buttonWrapper.find('button.button-weather__orange').simulate('click', { preventDefault: () => { } }));
        
    })

    test('return result response.json', () => {
        fetchMock.mockResponseOnce(JSON.stringify(['Riga', 'Latvia', REACT_APP_WEATHER_APPID]));
        const onResponse = jest.fn();
        const onError = jest.fn();

        return api('Riga', 'Latvia', REACT_APP_WEATHER_APPID)
            .then(onResponse)
            .catch(onError)
            .finally(() => {
                expect(onResponse).toHaveBeenCalled()
                expect(onError).not.toHaveBeenCalled();

                expect(onResponse.mock.calls[0][0]).toEqual(['Riga', 'Latvia', REACT_APP_WEATHER_APPID]);
            })
    });

    test('should return weather component initial state', () => {
        expect(new Weather().state).toEqual({
            temp: '',
            maxTemp: '',
            minTemp: '',
            main: '',
            city: 'Riga',
            cityToShow: '',
            countryToShow: '',
            country: 'LV',
            showWeather: false,
            showError: false
        })
    });

    test('Weather component renders correctly', () => {
        const tree = renderer
            .create(<Weather />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('render 2 input fields and 1 button GET WEATHER', () => {
        expect(wrapper.find('input').length).toBe(2);
        expect(wrapper.find('button').length).toBe(1);

    });

    test('updates first input field "city" ', () => {
        const fieldName = 'city';

        wrapper.find('[name="city"]').at(0).simulate('change', {
            target: {
                name: fieldName,
                value: 'Paris'
            }
        });
        expect(wrapper.state().city).toEqual('Paris');
    });

    test('updates second input field "country" ', () => {
        const fieldName = 'country';

        wrapper.find('[name="country"]').at(1).simulate('change', {
            target: {
                name: fieldName,
                value: 'France'
            }
        });
        expect(wrapper.state().country).toEqual('France');
    });

    test('fetches successfully data from OpenWeather API', async () => {

        axios.get.mockImplementationOnce(() => Promise.resolve(data));
        await expect(fetchData('Riga', 'Latvia', REACT_APP_WEATHER_APPID)).resolves.toEqual(data);

        wrapper.setState({ showWeather: true });
        expect(wrapper.state().showWeather).toEqual(true);

        expect(axios.get).toHaveBeenCalledTimes(1);

    });

    test("checks if fields rendered correctly", () => {

        const data = {
            result: {
                temp: '10',
                minTemp: '6',
                maxTemp: '12',
                description: 'rain'
            }
        };

        wrapper.setState({ temp: data.result.temp });
        expect(wrapper.find('span').at(1).text()).toEqual("10");

        wrapper.setState({ minTemp: data.result.minTemp });
        expect(wrapper.find('span').at(2).text()).toEqual("6");

        wrapper.setState({ maxTemp: data.result.maxTemp });
        expect(wrapper.find('span').at(3).text()).toEqual("12");

        wrapper.setState({ main: data.result.description });
        expect(wrapper.find('h3').at(3).text()).toEqual("rain");
    })

    test('fetches erroneously data from OpenWeather API', async () => {
        const errorMsg = 'Network error';
        axios.get.mockImplementationOnce(() => Promise.reject(new Error(errorMsg)));

        await expect(fetchData('Riga', 'Latvia', REACT_APP_WEATHER_APPID)).rejects.toThrow(errorMsg);
    });
    test('show error when search is not successfull', () => {
        wrapper.setState({ showError: true });

        expect(wrapper.find('div.weather-error').text()).toEqual('Incorrect City or Country');
    });
    test('show no error when search is successfull', () => {
        wrapper.setState({ showError: false });

        expect(wrapper.find('div.weather-error')).toHaveLength(0);
    });

});





