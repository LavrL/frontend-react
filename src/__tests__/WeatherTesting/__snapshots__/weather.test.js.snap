// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather testing should check if Weather component renders correctly 1`] = `
<div
  className="weather-block__center"
>
  <input
    className="weather__input-text"
    name="city"
    onChange={[Function]}
    type="text"
    value="Riga"
  />
  <input
    className="weather__input-text"
    name="country"
    onChange={[Function]}
    type="text"
    value="LV"
  />
  <button
    className="button-weather__orange"
    onClick={[Function]}
  >
    Get Weather
  </button>
  <div>
     
  </div>
</div>
`;
